creator-node/src/services/sync/secondarySyncFromPrimaryUtils.ts

```diff
@@ -0,0 +1,112 @@
+import type {
+  ForceResyncAuthParams,
+  ForceResyncConfig,
+  ForceResyncSigningData,
+  SyncRequestAxiosData
+} from '../stateMachineManager/stateReconciliation/types'
+
+const _ = require('lodash')
+
+const { logger: genericLogger } = require('../../logging')
+const config = require('../../config')
+const ContentNodeInfoManager = require('../stateMachineManager/ContentNodeInfoManager')
+const {
+  recoverWallet,
+  generateTimestampAndSignature
+} = require('../../apiSigning')
+
+const generateDataForSignatureRecovery = (body: SyncRequestAxiosData) => {
+  const { wallet, creator_node_endpoint, sync_type, immediate } = body
+  return {
+    wallet,
+    creator_node_endpoint,
+    sync_type,
+    immediate
+  }
+}
+
+// Function to sign sync data. This is used to determine whether or not to `forceResync`
+const signSyncData = (syncData: ForceResyncSigningData) => {
+  return generateTimestampAndSignature(
+    syncData,
+    config.get('delegatePrivateKey')
+  )
+}
+
+// Derive the Content Node delegate wallet from the data, signature, and timestamp
+const recoverWalletFromSyncData = ({
+  data,
+  timestamp,
+  signature
+}: ForceResyncAuthParams) => {
+  return recoverWallet({ ...data, timestamp }, signature)
+}
+
+/**
+ * Checks to see if the host requesting the sync is the primary of the observed user
+ * @param {Object} param
+ * @param {Object} param.apiSigning object containing the data, signaqture, and timestamp for signature verification
+ * @param {Object} param.libs the libs instance
+ * @param {string} param.wallet the observed user's wallet
+ * @param {boolean} param.forceResync flag from the request to force resync (i.e. clearing old user state) or not
+ * @param {Object} param.logger log object
+ * @param {Object} param.logContext object of log context. used when this sync job is enqueued
+ * @returns true or false, depending on the request flag and whether the requester host is the primary of the user
+ */
+const shouldForceResync = async (forceResyncConfig: ForceResyncConfig) => {
+  if (!forceResyncConfig) return false
+
+  let { apiSigning, wallet, forceResync, libs, logContext, logger } =
+    forceResyncConfig
+
+  if (logContext) {
+    logger = genericLogger.child(logContext)
+  } else if (!logger) {
+    logger = genericLogger
+  }
+
+  logger.debug(
+    `Checking shouldForceResync: wallet=${wallet} forceResync=${forceResync}`
+  )
+
+  if (!forceResync || !apiSigning) {
+    return false
+  }
+
+  const { data, timestamp, signature } = apiSigning
+  if (!data || !timestamp || !signature) {
+    return false
+  }
+
+  // Derive the Content Node delegate wallet from the data, signature, and timestamp
+  const recoveredPrimaryWallet = recoverWalletFromSyncData({
+    data,
+    timestamp,
+    signature
+  })
+
+  try {
+    // Get the delegate wallet from the primary of the observed user
+    const userPrimaryId = (await libs.User.getUsers(1, 0, null, wallet))[0]
```

Comment:
i think decent bit of duplicated code here from inside `ensurePrimaryMiddleware`? ideally we would refactor & reuse

Reply:
check out https://github.com/AudiusProject/audius-protocol/blob/vg-force-resync-only-on-primary/creator-node/src/middlewares.js#L751 for spID and delegateWallet recovery

Reply:
what part of this is duplicated code? could you send the snippet of code we should use here instead? I read through that middleware and it was too complex for our usage here

Reply:
so the entire `getCreatorNodeEndpoints` function is potentially reusable here - agreed it is much more complex and i can see why you chose not to use it. but it does do a lot more things to minimize errors e.g. retries

also just logically `ensurePrimaryMiddleware` and this function which generalizes to `ensureCallerIsPrimaryMiddleware` do very similar things

its fine to not do that work now but in the spirit of minimizing added code, would like to eventually consolidate a lot of this logic

Reply:
I didnt do the chain call because we only have context of the user wallet here in this flow. I would need to find the userId first, and then do the chain call to find the primary. I think it's okay to get the info from discovery for now

yeah we should probably simplify the ensurePrimaryMiddleware logic cus it's manually handling retries when we could be using the asyncRetry util fn \ has repeated code

Reply:
agreed on both